### ISSUE TITLE
POMDPs.stateindex is ambiguous

### ISSUE THREAD

**Issue by amrods** [JuliaPOMDP/POMDPs.jl#223]
It seems that `POMDPModelTools` is automatically loaded with `using POMDPs` creating ambiguities when implementing `stateindex`. See the notebook at https://github.com/amrods/pomdps/blob/master/mdps-Copy1.ipynb

```
MethodError: POMDPs.stateindex(::SimpleOG{Int64,Float64,Array{Float64,2},Array{Float64,3}}, ::Int64) is ambiguous. Candidates:
  stateindex(simpleog::SimpleOG, s) in Main at In[6]:3
  stateindex(mdp::Union{MDP, POMDP}, s::Int64) in POMDPModelTools at /Users/amrods/.julia/packages/POMDPModelTools/eHEjm/src/convenient_implementations.jl:28
Possible fix, define
  stateindex(::SimpleOG, ::Int64)
```

**Comment by amrods** [comment]
That makes `@requirements_info` throw a failure
```
For solve(::ValueIterationSolver, ::Union{MDP,POMDP}):
  [✔] discount(::SimpleOG)
  [✔] n_states(::SimpleOG)
  [✔] n_actions(::SimpleOG)
  [✔] transition(::SimpleOG, ::Float64, ::Float64)
  [✔] reward(::SimpleOG, ::Float64, ::Float64, ::Float64)
  [X] stateindex(::SimpleOG, ::Float64)
  [✔] actionindex(::SimpleOG, ::Float64)
  [✔] actions(::SimpleOG, ::Float64)
  [✔] support(::Array)
  [✔] pdf(::Array, ::Float64)
For ordered_states(::Union{MDP,POMDP}) (in solve(::ValueIterationSolver, ::Union{MDP,POMDP})):
  [✔] states(::SimpleOG)
For ordered_actions(::Union{MDP,POMDP}) (in solve(::ValueIterationSolver, ::Union{MDP,POMDP})):
  [✔] actions(::SimpleOG)
Note: Missing methods are often due to incorrect importing. You must explicitly import POMDPs functions to add new methods.
```
and still the ambiguity persists.
```
MethodError: POMDPs.actionindex(::SimpleOG{Int64,Float64,Array{Float64,2},Array{Float64,3}}, ::Int64) is ambiguous. Candidates:
  actionindex(simpleog::SimpleOG, a) in Main at In[7]:3
  actionindex(mdp::Union{MDP, POMDP}, a::Int64) in POMDPModelTools at /Users/amrods/.julia/packages/POMDPModelTools/eHEjm/src/convenient_implementations.jl:29
Possible fix, define
  actionindex(::SimpleOG, ::Int64)
```

**Comment by MaximeBouton** [comment]
Sorry I guess in my answer above I messed up the parametrization, based on your MDP definition it should probably be `T <: Real`, I edited the previous reply. 
The ambiguity is raised for a different function `actionindex` instead of `stateindex`.

**Comment by amrods** [comment]
I changed `actionindex` and `stateindex` and still I'm getting the same ambiguity:
```
MethodError: POMDPs.stateindex(::SimpleOG{Int64,Float64,Array{Float64,2},Array{Float64,3}}, ::Int64) is ambiguous. Candidates:
  stateindex(simpleog::SimpleOG, s::T) where T<:Real in Main at In[6]:3
  stateindex(mdp::Union{MDP, POMDP}, s::Int64) in POMDPModelTools at /Users/amrods/.julia/packages/POMDPModelTools/eHEjm/src/convenient_implementations.jl:28
Possible fix, define
  stateindex(::SimpleOG, ::Int64)
```

**Comment by zsunberg** [comment]
Hi @amrods, thanks for reporting this - it sounds quite frustrating. I do not understand why Maxime's solution didn't work.

In the short term, if you actually define `POMDPs.stateindex(::SimpleOG, s::Int64) = s+1` in addition to `stateindex(::SimpleOG, s::Any)`, I think it will work.

In the long term, I don't know what the solution is... maybe we need to just get rid of that implementation?

**Comment by MaximeBouton** [comment]
What do you think of: 
```julia 
stateindex(mdp, s::Bool) = s+1
stateindex(mdp, s::Int) = s
```
This would certainly avoid the ambiguity but it does not feel super right.
Getting rid of those is an option as well.

**Comment by zsunberg** [comment]
I don't think that would actually fix the ambiguity error. Here is a minimal snippet to test ideas:
```julia
using POMDPs

# The line below is what would be in POMDPModels
POMDPs.stateindex(::Any, s::Int) = s

struct M{T} <: MDP{T, T} end

m = M{Int}()

POMDPs.stateindex(::M, s::S) where S<:Real = s

@show stateindex(m, 1)
# ambiguity error
```

**Comment by zsunberg** [comment]
to be clear, it doesn't have anything to do with the parameterization of `M`
```julia
using POMDPs

# The line below is what would be in POMDPModels
POMDPs.stateindex(::Union{POMDP, MDP}, s::Int) = s

struct M <: MDP{Int, Int} end

m = M()

POMDPs.stateindex(::M, s::S) where S<:Real = s

@show stateindex(m, 1)
# ambiguity error
```
also doesn't work

**Comment by zsunberg** [comment]
maybe it's not too much to ask people to write that one extra `stateindex(::MyMDP, ::Int)` method.

**Comment by zsunberg** [comment]
Ok, I am ok with making that change. We just need to update POMDPModels first.

**Comment by MaximeBouton** [comment]
We need to update the tutorials as well.

**Comment by MaximeBouton** [comment]
Are we removing all of them? 
I am actually unsure if it is a wise move anymore.

**Comment by zsunberg** [comment]
We should figure out which ones will cause ambiguity errors.

In previous versions of Julia, `stateindex(::M, s::S) where S` as you suggested above would have worked fine. I am not sure why it doesn't. Maybe it is a bug and they are going to fix it.

**Comment by MaximeBouton** [comment]
I spoke in person with @rejuvyesh , we came to conclusion that it would be clearer to remove them and instead add a note in the docs to show a simple way to implement `stateindex` (and friends) for Bool and Int representations. 

We could probably keep these: 
```julia 
rand(rng::AbstractRNG, t::Tuple{Bool, Bool}) = rand(rng, Bool)
rand(t::Tuple{Bool, Bool}) = rand(Bool)
support(s::AbstractVector) = s
support(s::Tuple) = s
support(r::AbstractRange) = r
support(g::Base.Generator) = g
```

**Comment by MaximeBouton** [comment]
It should be fixed with JuliaPOMDP/POMDPModelTools#12
